collector/lib/CollectorService.cpp

```diff
@@ -148,8 +150,31 @@ void CollectorService::RunForever() {
   sysdig_.CleanUp();
 }
 
-bool CollectorService::InitKernel() {
-  return sysdig_.InitKernel(config_);
+bool CollectorService::InitKernel(const std::string& GRPCServer, const std::vector<DriverCandidate>& candidates) {
+  auto& startup_diagnostics = StartupDiagnostics::GetInstance();
+
+  for (const auto& candidate : candidates) {
+    if (!GetKernelObject(GRPCServer, config_.TLSConfiguration(), candidate, config_.CurlVerbose())) {
+      CLOG(WARNING) << "No suitable kernel object downloaded for " << candidate.GetName();
+      startup_diagnostics.DriverUnavailable(candidate.GetName());
+      continue;
+    }
+
+    startup_diagnostics.DriverAvailable(candidate.GetName());
+
+    if (sysdig_.InitKernel(config_, candidate)) {
+      startup_diagnostics.DriverSuccess(candidate.GetName());
+      return true;
+    } else if (candidate.GetCollectionMethod() == KERNEL_MODULE) {
+      // Kernel module drops capabilities, so subsequent attempts could fail
```

Comment:
Could we ease this a little by enforcing that kernel module is always the last candidate? Probs wouldn't affect it if we were trying multiple kernel modules, and also we'll be removing kernel modules soon anyway... this is probably fine.

Reply:
I disagree. The whole logic may change multiple time, and it's just nice to have some invariants verified.